### PR TITLE
[PVR] Fix CPVRTimers::GetRecordingTimerForRecording to respect timer'…

### LINKED ISSUE #15421: PVR: Recordings window no longer shows recording icon for in-progress recordings
### Describe the bug
Here is a clear and concise description of what the problem is:

In Krypton, the PVR Recordings window used to show a recording (red dot) icon next to any recordings that were currently in-progress. This is no longer the case in Kodi Leia 18.0

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

In the PVR Recordings window, the recording icon (red dot) should show next to any items that are currently recording.

## Actual Behavior

For an item that is recording, it just shows the normal icon for an unwatched, finished recording.

## Possible Fix

From what I've found, when in the PVR Recordings window, it seems that the ListItem.IsRecording boolean condition returns false for shows that are currently recording.

### To Reproduce

1. Start a recording on a TV Show
2. Navigate to the PVR Recordings window
3. See whether the icon next to the recording shows as the recording in progress icon (windows/pvr/record.png)

### Debuglog

The debuglog can be found here:

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [X] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: OSMC
 - Kodi version: 18.0

 Running pvr.hts against TVHeadend server version 4.2.5

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by ksooo**
…s start and stop padding time + optimize to check epg event uids if present.

Fixes #15421 

@Jalle19 mind doing a quick review?

### CODE REVIEW

File: xbmc/pvr/timers/PVRTimers.cpp [hunk 1]
```diff
@@ -796,11 +796,21 @@ CPVRTimerInfoTagPtr CPVRTimers::GetRecordingTimerForRecording(const CPVRRecordin
...
-          timersEntry->m_iClientChannelUid == recording.ChannelUid() &&
-          timersEntry->StartAsUTC() <= recording.RecordingTimeAsUTC() &&
-          timersEntry->EndAsUTC() >= recording.EndTimeAsUTC())
+          timersEntry->m_iClientChannelUid == recording.ChannelUid())
       {
-        return timersEntry;
+        // match epg event uids, if available
+        if (timersEntry->UniqueBroadcastID() == recording.BroadcastUid())
+        {
+          if (timersEntry->UniqueBroadcastID() != EPG_TAG_INVALID_UID)
```
Comment (Jalle19) [inline xbmc/pvr/timers/PVRTimers.cpp]: These two if statements could be combined

**Review by Jalle19 — APPROVED** [review]

File: xbmc/pvr/timers/PVRTimers.cpp (hunk 1 above)
Comment (ksooo) [reply xbmc/pvr/timers/PVRTimers.cpp]: Yep. I will change that.

**Comment by ksooo** [comment]
jenkins build this please

**Comment by linknetx** [comment]
Did this fix find it's way into the last nightly build KodiSetup-20190206-7001c652-master-x64 ?

If it did then this hasn't fixed the issue for this setup:

Kodi 18.0
TVMosaic 1.0.0 17051
TVMosaic PVR addon 4.7.0

The current recording doesn't show up in the recordings list.

**Comment by ksooo** [comment]
Yes the fix went in and it definitely fixes this for me  (pvr.hts). Maybe we've missed an edge case? Maybe tvmosaic behaving different wrt start and end times reported for recordings and timers... That could make a difference.

IIRC, you are a developer. How about debugging this by yourself?

**Comment by linknetx** [comment]
Am I a developer, afraid not. I just do some coding in my spare time for my own projects normally in C#. My skills in C++ are somewhat limited and that particular segment of code is a stretch too far!

**Comment by ksooo** [comment]
I could provide you with a special build with additional locking to try to identify the problem this way? Would you have some spare time for this? You are using Windows, right?

**Comment by linknetx** [comment]
I've got plenty of spare time and I'm using Windows 10 x64, so please go 
ahead.

------ Original Message ------

**Comment by linknetx** [comment]
@ksooo 

By hit and miss it works with this part of the code commented out:

`       /* && timersEntry->m_iClientChannelUid == recording.ChannelUid() */`

in CPVRTimers::GetRecordingTimerForRecording:

`      if (timersEntry->IsRecording() &&
          !timersEntry->IsTimerRule() &&
          timersEntry->m_iClientId == recording.ClientID() /* &&
          timersEntry->m_iClientChannelUid == recording.ChannelUid()*/ )`

Not sure why but hope that this helps.

**Comment by ksooo** [comment]
Let's see what comes out of the extened logging session. Here is the link for the test build.

http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20190207-1b9cbdf4-pvr-fix-isrecording-x64.exe

Please enable PVR component logging, start a recording, goto recordings window (verify that the red dot is not where it should be), and post the debug log, then.

**Comment by linknetx** [comment]
OK, done that, and can confirm that the red dot is not showing in recordings list on the recording item.

Here is the log:

[kodi.log](https://github.com/xbmc/xbmc/files/2840343/kodi.log)

**Comment by ksooo** [comment]
Sorry, you have not enabled debug logging. So, log contains nothing of interest.

Goto settings -> system -> logging. enable logging AND component logging for pvr

**Comment by linknetx** [comment]
OK, try again:

[kodi.log](https://github.com/xbmc/xbmc/files/2840499/kodi.log)

**Comment by ksooo** [comment]
OK, I now know what goes wrong. Unfortunately, this cannot work if the pvr addon (pvr.dvblink in your case) does not supply the channel uid with its recordings. This is an API change rquested some years(!) ago, btw.

=> https://github.com/kodi-pvr/pvr.dvblink/blob/master/src/DVBLinkClient.cpp#L1456

You should file an issue for pvr.dvblink. Please note that it is no bug that the adodn does not implement this; the channel uid is not mandatory. But if this is not done, some (non crucial) Kodi features, like the "red dot" simply cannot work. So, you should kindly ask the adon dev to implement this, pointing out what the consequence of the mssing feature is.

**Comment by linknetx** [comment]
OK, thanks for your help,

Just so that I'm clear about this does this need fixing in both the pvr.dvblink server and the dvb.link client?

**Comment by ksooo** [comment]
The addon needs to be fixed. 
Not sure what you mean with pvr.dvblink client - if you mean Kodi core code changes, then the answer is no.

**Comment by linknetx** [comment]
I understand that the addon needs to be fixed, so would this also 
require corresponding changes on the dbvlink/tvmosaic backend server ?

------ Original Message ------

**Comment by ksooo** [comment]
> so would this also require corresponding changes on the dbvlink/tvmosaic backend server

 Sorry, don't know .

**Comment by linknetx** [comment]
OK, thanks again, I'll refer it to the pvr.dvblink developers.

**Comment by linknetx** [comment]
@ksooo 

I've put in a request for DVBLogic to implement the ChannelUid so hopefully this should get fixed eventually. 

Since the ChannelUid is optional and disables one of the features of Kodi, namely the red dot recording icon, would it be better to have a general solution where the feature works without the optional ChannelUid but uses it when available.

I'm sure you must have considered matching channel names for the case where the ChannelUid is invalid and rejected it for a valid reason:

if (timersEntry->IsRecording() &&
    !timersEntry->IsTimerRule() &&
    timersEntry->m_iClientId == recording.ClientID() &&
    recording.ChannelUid() == PVR_CHANNEL_INVALID_UID ?
    timersEntry->ChannelName() == recording.m_strChannelName :
    timersEntry->m_iClientChannelUid == recording.ChannelUid())

One reason for rejecting it would be that no channel names are available (is that possible?) but that can be checked for:

if (timersEntry->IsRecording() &&
    !timersEntry->IsTimerRule() &&
    timersEntry->m_iClientId == recording.ClientID() &&
    (recording.ChannelUid() == PVR_CHANNEL_INVALID_UID &&
        timersEntry->ChannelName().size() > 0 &&
        recording.m_strChannelName.size() > 0) ?
    timersEntry->ChannelName() == recording.m_strChannelName :
    timersEntry->m_iClientChannelUid == recording.ChannelUid())

Another reason for rejecting it would be if two (or more) channels had identical names, I suppose that's possible, but they would both (or all) have to be currently recording for it to cause a problem.

I think that this minor issue would be better than the current situation where the ChannelUid isn't available. Which is that is a currently recording item shows the delete option in the context menu, which doesn't work and is confusing, rather than the correct stop recording option.

By the way I've tested both of these and they work perfectly.

**Comment by ksooo** [comment]
> Since the ChannelUid is optional and disables one of the features of Kodi, namely the red dot recording icon,

 To be acuurate, it prevents the red dot only in the recordings window, it works everywhere else.

 So, it is just a very minor glitch, if at all. As I said, no essential functuonality is bound to presense of channel uid.

 Right approach is not to hack around with comparing channel names and stuff, but to make channeluid mandatory for PVR_RECORDING. Especially, as I'm convinced that it is a now brainer for pvr addon devs to add support for this (they already have that information in the addon).

 Problem is, that some addon devs are quite lazy (err, don't have enough free time) and only implement mandatory stuff or do something if users complain.

**Comment by linknetx** [comment]
>Right approach is not to hack around with comparing channel names and stuff, but to make channeluid mandatory for PVR_RECORDING.

That's a better solution, can that be implemented?

**Comment by ksooo** [comment]
> That's a better solution, can that be implemented?

 Actual work needs to be done by the addon maintainers.

 All I can (need to) do on the kodi side is updating the api documentation (optional -> mandatory) and announcing the API change for v19, making it a bug if the addon does not support PVR_RECORDING.channeluid.

**Comment by FernetMenta** [comment]
@ksooo a backend can record some program on a channel that is not synchronized with Kodi. Would this still be possible if PVR_RECORDING.channeluid was mandatory?

**Comment by ksooo** [comment]
This all a bitt messy. I just discovered that PVR_TIMER.channeluid is also optional. So, we would also need to make PVR_TIMER.channeluid mandatory. :-(

PVR_RECORDINGs (in difference to PVR_TIMERs) have no state. That's imo the root cause of the "red dot" problem.

To determine a recording's "recording in progress" state (to be able to support the red dot, for instance), one must search for the timer that belongs to a recording . And for a 100% match you need the channel uid both in the timer and the recording, which will obviously not work 100% as long as channel uids are optional.  :-/

Please note that it is not enough to only check recording's time and duration against 'now'. Only the state of the related timer currently is a safe indicator that the backend actually is recording this. That's the reason I implemented it the way it is today.

Solution would be another mandatory API change - to inroduce PVR_RECORDING.state (or an in-progress bool).

**Comment by linknetx** [comment]
>To be accurate, it prevents the red dot only in the recordings window; the dot works everywhere else.

It also doesn't work on the recent recordings on the home screen

>So, it is just a very minor glitch, if at all. As I said, no essential functionality is bound to presence of channel uid.

The lack of a red dot is not really the issue, the problem arises when you attempt to delete a recording from the recordings list which is still recording. The context menu gives the delete option which does nothing and since you are not aware that it is still recording, no red dot, it's somewhat confusing. It then involves finding it in the timers list, stopping recording then back to the recordings list to delete it.

**Comment by ksooo** [comment]
You are right, i overlooked the other places where the problem occurs. Thanks for mentioning.

Tell the dvblogic guys to implement the one liner and we're fine for all pvr.dvblink users.

This pragmatic approach has low effort and high gain and is not a hack.

**Comment by ksooo** [comment]
> The context menu gives the delete option which does nothing 

 Not sure this another unrelated problem  with the addon, as this works fine with other addons.

**Comment by linknetx** [comment]
> Not sure this another unrelated problem with the addon, as this works fine with other addons.

This is only if the recording is in progress, it won't delete it whilst still recording. For all completed recordings it deletes correctly.

**Comment by ksooo** [comment]
> This is only if the recording is in progress, it won't delete it whilst still recording. For all completed recordings it deletes correctly.

 I understood this in the first place. And as I said, this works with other pvr addons, thus I tend to say, open a ticket for pvr.dvblink (or start using another, better supported pvr addon). ;-)

**Comment by linknetx** [comment]
If pvr.dvblink implement the ChannelUid then I will get the red dot in the recordings list and the context menu will give the 'Stop recording' option which works :-)

By better supported I take it you mean TVHeadend, as far as I understand that means a a switch to Linux on the backend, which I don't really want.

Now if TVHeadend could be ported to Windows !!!!

**Comment by ksooo** [comment]
> If pvr.dvblink implement the ChannelUid then I will get the red dot in the recordings list 

 Yes.

> and the context menu will give the 'Stop recording' option which works

 Most probably not. I tried to explain that most probably another fix in pvr.dvblink is needed for this. We're facing two different issues here, thus two fixes are needed.

**Comment by linknetx** [comment]
If I implement the hack, which I suggested a few posts previously, to match the channels via Channel Name rather than Channel Uid then I get the red dot and I also get the context menu with the 'Stop Recording' option which works fine. Had it running for a few days and it works every time :-)

**Comment by ksooo** [comment]
Channel names can change at any time and nobody guarantees that record's and timer's channel name have to match. They are just strings, API does not require that consistent channel names must be used across timers and recordings. Thus, sorry, cannot accept your hack.

And now, I will end this discussion from my end. Please understand that my time is limited. No bad blood involved, just not enough time.

**Comment by linknetx** [comment]
No problem whatsoever, I understand entirely and thanks for your help, no reply needed :-)

**Pull request merged**